crates/wasi/src/preview2/tcp.rs

```diff
@@ -103,23 +107,76 @@ impl HostInputStream for HostTcpSocketInner {
 
 const SOCKET_READY_SIZE: usize = 1024 * 1024 * 1024;
 
+pub(crate) struct TcpWriteStream {
+    stream: HostTcpSocketInner,
+    write_handle: Option<tokio::task::JoinHandle<anyhow::Result<()>>>,
+}
+
+impl TcpWriteStream {
+    pub(crate) fn new(stream: HostTcpSocketInner) -> Self {
+        Self {
+            stream,
+            write_handle: None,
+        }
+    }
+
+    /// Write `bytes` in a background task, remembering the task handle for use in a future call to
+    /// `write_ready`
+    fn background_write(&mut self, mut bytes: bytes::Bytes) {
+        assert!(self.write_handle.is_none());
+
+        let stream = self.stream.clone();
+        self.write_handle.replace(tokio::spawn(async move {
+            while !bytes.is_empty() {
+                stream.tcp_socket().writable().await?;
+                let n = stream.tcp_socket().try_write(&bytes)?;
+                let _ = bytes.split_to(n);
+            }
+
+            Ok(())
+        }));
+    }
+}
+
 #[async_trait::async_trait]
-impl HostOutputStream for HostTcpSocketInner {
+impl HostOutputStream for TcpWriteStream {
     fn write(&mut self, mut bytes: bytes::Bytes) -> anyhow::Result<Option<WriteReadiness>> {
         while !bytes.is_empty() {
-            let n = self.stream.try_write(&bytes)?;
-            let _ = bytes.split_to(n);
+            match self.stream.tcp_socket().try_write(&bytes) {
+                Ok(n) => {
+                    let _ = bytes.split_to(n);
+                }
+
+                Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => {
+                    self.background_write(bytes);
+
+                    // return None to indicate that the write is in progress, and that waiting is
+                    // needed for further writes.
+                    return Ok(None);
+                }
+
+                Err(e) => return Err(e.into()),
```

Comment:
This may be preexisting, but I think this means that I/O errors turn into trapping errors? Or at least I think that the `Err` here gets turned into a trap, but I could easily be misremembering

Reply:
The error handling scheme gets fixed in my branch to explicitly call out operation failed vs traps.

Reply:
I've rebased on Pat's branch, and reworked the error handling to categorize error results as traps or runtime errors.